### DISCUSSION TITLE
sessions table not pruning?

### DISCUSSION THREAD

**Discussion by sts-ryan-holton** [laravel/framework#52123] [Q&A]
### Laravel Version

11.15.0

### PHP Version

8.3.7

### Database Driver & Version

MySQL 8

### Description

I'm using the sessions table for authentication. I only have a user or two in my database, yet have over 9,000 sessions stored in the sessions table. I'm using Laravel Livewire and everything in one system.

Really the sessions table should be pruned

### Steps To Reproduce

env config:

```
SESSION_DRIVER=database
SESSION_LIFETIME=1440
SESSION_ENCRYPT=false
SESSION_PATH=/
SESSION_DOMAIN=null
```

**Comment by rodrigopedra** [comment]
Session data is pruned depending on a random number generated at each request. 

This mechanic is referred as lottery. It is configured on your app's `./config/session.php` file:

https://github.com/laravel/laravel/blob/4ef5e2f89e987f84b33b62f79e96485dcaa8f209/config/session.php#L106-L117

If your app has a very low request count, you can try increasing the odds of session data being pruned (aka garbage collected).

As an alternative, you can write an artisan command that runs this garbage collection mechanics, and use the command scheduler to run it periodically.

**Comment by rodrigopedra** [comment]
This should help you out

```php
<?php // ./routes/console.php

use Illuminate\Support\Facades\Artisan;

Artisan::command('session:gc', function () {
    session()->getHandler()->gc(60 * config('session.lifetime'));
})->twiceDaily();
```

**Comment by sts-ryan-holton** [comment]
@rodrigopedra My application gets around 750 - 1,000 requests per minute with each request ranging from maybe a few ms to seconds.

I'm slightly confused on the sessions table then because I thought this table would exclusively be for "users" of the application, e,g: those that log in to the system.

I have some publically accessible API endpoints that are consumed but other systems that are no-auth endpoints, it would appear that these too enter the sessions table but are never cleared?

**Comment by rodrigopedra** [comment]
Every request that goes under the `web` middleware group will have a session created.

If your public facing site does not need any kind of session data, you can create a new middleware group and apply to those routes.

Regarding the API endpoints, those should not have any session attached. Run:

```bash
php artisan route:list -v
```

And check which middleware are applied to these routes. The `-v` parameter is what tells the `route:list` command to show each route's middleware, don't omit it.

Also, only define API endpoints on your `./routes/api.php` file.

**Comment by sts-ryan-holton** [comment]
API routes are only defined in my API routes file. These are the public ones. I've implemented that session garbage handler command which I'll deploy next week. Just to clarify the purposes of doing the `60 * config('session.lifetime')`?

My session lifetime variable is 1440, I'm presuming I could reduce the 60 down to something like 2 or 3?

**Comment by rodrigopedra** [comment]
The `gc()` methods expects the session lifetime in seconds. 

In the configuration file, you define it in minutes.

The `60 * config('session.lifetime')` just converts minutes to seconds.

Internally, it will delete sessions that are stale for more than that threshold.

https://github.com/laravel/framework/blob/f5f9383af919ef74b00eacb23b4a88fcf390743c/src/Illuminate/Session/DatabaseSessionHandler.php#L279-L282

**Comment by rodrigopedra** [comment]
That is the same that the `StartSession` middleware uses when calling the garbage collection mechanism:

https://github.com/laravel/framework/blob/f5f9383af919ef74b00eacb23b4a88fcf390743c/src/Illuminate/Session/Middleware/StartSession.php#L170-L180

https://github.com/laravel/framework/blob/f5f9383af919ef74b00eacb23b4a88fcf390743c/src/Illuminate/Session/Middleware/StartSession.php#L254-L257
